server.js

```diff
@@ -20,4 +35,89 @@ app.get("/", function (request, response) {
   response.sendFile(__dirname + "/index.html");
 });
 
-app.listen(process.env.PORT);
+// Get all messages
+app.get("/messages", function (request, response) {
+  response.status(200).json(messages);
+});
+
+// Get all messages by search term
+app.get("/messages/search", function (request, response) {
+  if (request.query.text) {
+    let filteredMessages = messages.filter((msg) =>
+      msg.text.includes(request.query.text)
+    );
+    if (filteredMessages.length == 0) {
+      response.status(400).json({ msg: "No matching results" });
+    } else {
+      response.status(200).json(filteredMessages);
+    }
+  } else {
+    response.status(400).json({ msg: "Please enter search term" });
+  }
+});
+
+// Get 10 latest messages
+app.get("/messages/latest", function (request, response) {
+  let latestMessages = messages.slice(-10).reverse();
+  response.status(200).json(latestMessages);
+});
+
+// Get one message by id
+app.get("/messages/:id", function (request, response) {
+  let message = messages.find((msg) => msg.id == request.params.id);
+  if (message) {
+    response.status(200).json(message);
+  } else {
+    response.status(400).json({
+      msg: "No message with the Id '" + request.params.id + "' is found",
+    });
+  }
+});
+
+// Create new message
+app.post("/messages", urlencodedParser, function (request, response) {
+  if (!request.body.from || !request.body.text) {
+    response.status(400).json({ msg: "Please enter all fields" });
+  }
+
+  const newMessage = {
+    id: uuid.v4(),
+    from: request.body.from,
+    text: request.body.text,
+    timeSent: new Date(),
+  };
+
+  messages.push(newMessage);
+  response.status(200).json(messages);
+});
+
+// Update one message by id
+app.put("/messages/:id", function (request, response) {
+  let msgIndex = messages.findIndex((msg) => msg.id == request.params.id);
+  
+  isInvalidId(request.params.id, msgIndex, response);
+
+  if (request.body.text) {
+    messages[msgIndex].text = request.body.text;
+  }
+  if (request.body.from) {
+    messages[msgIndex].from = request.body.from;
+  }
+
+  response.status(200).json(messages);
```

Comment:
This is fine, but it looks more like a patch endpoint than a put endpoint to me. In a put endpoint you should replace the entire document, not just edit the fields that are different.